rollup/ccc/logger.go

```diff
@@ -0,0 +1,859 @@
+// resource usage of associated subcircuit
+func (l *Logger) logPrecompileAccess(to common.Address, inputLen uint64, inputFn func(int64, int64) []byte) {
+	l.logCopy(inputLen)
+	var outputLen uint64
+	switch to {
+	case common.BytesToAddress([]byte{1}): // &ecrecover{},
+		l.sigCount++
+		l.keccakUsage += computeKeccakRows(64)
+		outputLen = 32
+	case common.BytesToAddress([]byte{2}): // &sha256hash{},
+		l.logSha256(inputLen)
+		outputLen = 32
+	case common.BytesToAddress([]byte{3}): // &ripemd160hashDisabled{},
+	case common.BytesToAddress([]byte{4}): // &dataCopy{},
+		outputLen = inputLen
+	case common.BytesToAddress([]byte{5}): // &bigModExp{eip2565: true},
+		const rowsPerModExpCall = 39962
+		l.modExpUsage += rowsPerModExpCall
+		if inputLen >= 96 {
+			outputLen = new(big.Int).SetBytes(inputFn(64, 32)).Uint64() // mSize
+		}
+	case common.BytesToAddress([]byte{6}): // &bn256AddIstanbul{},
+		l.ecAddCount++
+		outputLen = 64
+	case common.BytesToAddress([]byte{7}): // &bn256ScalarMulIstanbul{},
+		l.ecMulCount++
+		outputLen = 64
+	case common.BytesToAddress([]byte{8}): // &bn256PairingIstanbul{},
+		l.ecPairingCount++
+		outputLen = 32
+	case common.BytesToAddress([]byte{9}): // &blake2FDisabled{},
+	}
+	l.logCopy(2 * outputLen)
+}
+
+// logCall logs call to a given address, regardless of the address being a precompile or not
+func (l *Logger) logCall(to common.Address, inputLen uint64, inputFn func(int64, int64) []byte) {
+	l.logBytecodeAccessAt(to)
+	l.logPrecompileAccess(to, inputLen, inputFn)
+}
+
+func (l *Logger) logCopy(len uint64) {
+	l.copyUsage += len * 2
+}
+
+func (l *Logger) logSha256(inputLen uint64) {
+	const blockRows = 2114
+	const blockSizeInBytes = 64
+	const minPaddingBytes = 9
+
+	numBlocks := (inputLen + minPaddingBytes + blockSizeInBytes - 1) / blockSizeInBytes
+	l.sha256Usage += numBlocks * blockRows
+}
+
+func (l *Logger) CaptureStart(env *vm.EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) {
+	l.currentEnv = env
+	l.isCreate = create
+	if !l.isCreate {
+		l.logCall(to, uint64(len(input)), func(argOffset, argLen int64) []byte {
+			return input[argOffset : argOffset+argLen]
+		})
+	} else {
+		l.logRawBytecode(input) // init bytecode
+	}
+
+	if !env.TxContext.IsL1MessageTx {
+		l.sigCount++
+		l.l2TxnsRlpSize += uint64(env.TxContext.TxSize)
+	}
+	l.keccakUsage += computeKeccakRows(uint64(env.TxContext.TxSize))
+	l.keccakUsage += computeKeccakRows(64) // ecrecover per txn
+}
+
+func (l *Logger) CaptureState(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, rData []byte, depth int, err error) {
+	if err != nil {
+		return
+	}
+
+	l.evmUsage += evmUsagePerOpCode[op]
+	l.stateUsage += stateUsagePerOpCode[op](scope, depth)
+
+	getInputFn := func(inputOffset int64) func(int64, int64) []byte {
+		return func(argOffset, argLen int64) []byte {
+			// it is safe to ignore the error, with current block gas limits (10M) it is not possible to go past
```

Comment:
At what gas limit would this become unsafe?

Reply:
~20M

Reply:
Got it. My only concern is that we'll increase gas limit in the future and forget about this. In that case, would we need to increase `memoryPadLimit` again?

Reply:
Yeah that would be a pain to deal with. Added error handling https://github.com/scroll-tech/go-ethereum/pull/982/commits/843f5fa12276c314b73ce5d424a57aae0641a56f